Modules/_csv.c

```diff
@@ -1342,75 +1281,98 @@ csv_writerow(PyObject *op, PyObject *seq)
             quoted = 1;
             break;
         case QUOTE_STRINGS:
-            quoted = PyUnicode_Check(field);
+            quoted = !is_none && PyUnicode_Check(field);
             break;
         case QUOTE_NOTNULL:
-            quoted = field != Py_None;
+            quoted = !is_none;
             break;
-        default:
+        default: /* Default QUOTE_MINIMAL or other to minimal behavior */
             quoted = 0;
             break;
         }
 
-        null_field = (field == Py_None);
-        if (PyUnicode_Check(field)) {
-            append_ok = join_append(self, field, quoted);
-            Py_DECREF(field);
+        if (is_none) {
+            /* None is NULL. */
+            str_field = NULL;
+        } else if (PyUnicode_Check(field)) {
+            str_field = Py_NewRef(field);
+        } else {
+            str_field = PyObject_Str(field);
+            if (str_field == NULL) {
+                Py_DECREF(field);
+                goto error;
+            }
         }
-        else if (null_field) {
-            append_ok = join_append(self, NULL, quoted);
-            Py_DECREF(field);
+
+        /* Single empty field special case */
+        if (field_count == 0) {
+            first_field_was_none = is_none;
+            if (is_none) {
+                first_field_was_empty_like = true;
+            } else {
+                Py_ssize_t len = str_field ? PyUnicode_GET_LENGTH(str_field) : 0;
+                first_field_was_empty_like = (len == 0);
```

Comment:
```suggestion
            first_field_was_empty_like = is_none || !PyUnicode_GET_LENGTH(str_field);
```

But rename the `is_none` variable. It's not possible to have `str_field == NULL` if `i s_none` is false.